Edits to Sigleton Pattern in SMSMessageParser

ricdezen opened the pull request:
After discussing with Cognolato and Crema the singleton pattern has been reintroduced.
Several stuff has been edited:
- Methods in SMSMessageParser take a Context parameter to access Preferences.
- Default strategy is standalone for easier construction since getInstance requires a Context.
- Added SMSParseStrategy extending MessageParseStrategy to make instantiation of class from Preferences much easier.
- Added Mockito and Robolectric for testing.
- Tests have been added, they weren't working, then they just started working by themselves and I am quite scared. Help.

I'm not entirely happy with the results because Context is always needed. Despite this, having the user be able to replace the default parsing strategy through a context is better than not letting him change it at all, it is very unlikely to have an access to our library outside of a Context.

ricdezen committed:
Merge pull request #1 from Cogno-IDU/master

Pull from Remote

ricdezen committed:
Merge pull request #2 from Cogno-IDU/master

Implemented delivery report (#14)

Commit:
Swapped arguments in MessageParseStrategy

They were reversed compared to the handler and it was bugging me immensely

Commit:
Removed getInstance from SMSMessageHandler

This class should not be singleton, should be an object pool or a normal class.

Cogno-IDU commented:
> I suggest switching over to an object pool

Uhm, I don't see the object pool...
The object pool is a design pattern where instead of creating objects yourself (with `new YourObject(...)` ) you use a singleton to get them (so `ObjectPool.getInstance().getObject()`) and you let the object pool handle creation of objects (and optimizations like instead of deleting them they can be kept alive and reused later)
[Here's a link](https://sourcemaking.com/design_patterns/object_pool) on how the object pool design pattern works.

I guess you (correctly) wanted to remove the singleton, and I get what's the benefit of removing it, but is this the way to do it or is there a better way?

RicDeZen commented:
I know what an Object Pool is, I simply haven't implemented yet it because I'd like to discuss it and other options beforehand

Cogno-IDU commented:
Oh alright, sorry!
Well, I don't see why having an object pool would be useful here. Since as you say it would be better to have objects parsing in different ways, with the object pool you would only get objects with the same parsing strategy or (at best) a "random" parsing strategy. 
This is because you basically create an object with a custom strategy by extending the Strategy class, but if you give it to the object pool so it can handle it instead of you, and another user gets an object in the objectpool, it might pass your strategy to the other user, instead of you.
Does it make sense?

Cogno-IDU commented:
What I would see more useful would just be, probably something like when you want to parse the message you directly pass also the strategy, and if you don't give one the default strategy is used.

RicDeZen commented:
Yeah that issue came up while I was trying to figure out a solution, same goes for passing the strategy directly in the method. I was also thinking we could add a method to the interface to detect if two strategies can be considered the same, kinda like equals (or just turn the strategy into an abstract class and use equals itself), this would be nice for strategies that don't have any mutable fields, you can just pass them once when you get the instance

CremaLuca commented:
So, let me get this right: you want to give the possibility to multiple modules of one app to have multiple strategies, right?
1. Are you sure that two modules accessing the same singleton (as a library, not a module) access to the same class? Did you test this? Because I think this library won't be used as a module but more like a build.gradle implementation (library) and I think that would lead to having two different sets of classes of the library but idk, must be somehow tested.
2. How would the BroadcastReceiver decide which strategy is correct when a message arrives?

I'm more prone to thinking that a more correct answer would be to somehow have multiple classes for each module using the library and that the answer is not ObjectPool.

RicDeZen commented:
I had found in the past as long as the class has the same package and name, the instance will be shared, an example is reported here: https://stackoverflow.com/a/21656109
I was thinking the BroadcastReceiver could just use the default strategy (either our default or user-set default) and then another strategy would be used in an upper level for further parsing

CremaLuca commented:
Confirmed with a test: if two classes of two modules were to do `SMSMessageHandler.getInstance.setStrategy` they'd override the other class's choice

RicDeZen commented:
Since it's not extremely urgent I'd like discuss this tomorrow IRL as Marco suggested so that other people can join us

CremaLuca commented:
What about a whole new set of classes on a "higher" level that can handle multiple instances of classes for different modules differentiating them by a code or by their package?
- we would need a class that extends `SMSMessage` with an identification
- the strategy "beneath" this level would have to parse messages and read their identification (from sms data)
- this level would have to handle multiple received listeners (at least one for each module)
- it would also have to handle multiple sent/delivered listeners

So, the current library would still be a singleton, and have a big adapter/wrapper around it. It's doable, but I don't think it's worth the cost right now.

RicDeZen commented:
Perhaps I wasn't clear with the title, my suggestion is to remove the singleton from SMSMessageHandler only, not SMSManager, If I understood your suggestion, it doesn't sound much different from simply parsing in the upper level, I don't really see the need to force a routing of the messages to a module when they are received, that's why I simply proposed to have an Object pool (it would actually be a Multiton but the line is kinda blurry) using the strategies themselves as keys for the instances

Commit:
Minimal fixes to SMSMessageHandler

- Removed private instance field
- Fixed spec in constructor

Commit:
Reverted to Singleton design pattern

Commit:
Merge branch 'master' of https://github.com/Cogno-IDU/smslibrary into hotfix/smshandler-strategy

# Conflicts:
#	smslibrary/src/main/java/com/eis/smslibrary/SMSHandler.java

Commit:
Added ability to save and read a default SMS parsing strategy

- Most SMSMessageParser methods take context as argument
- Added SMSParseStrategy as an interface to simplify use of reflection
- Added Robolectric to build.gradle for testing

Commit:
Fixed a TODO and some errors

- Constructor now reads from Preferences as expected.
- DefaultSMSParseStrategy is standalone to avoid having to pass a context to instantiate it.

CremaLuca commented:
Ok so, as discussed yesterday I'd say that you could keep the edits on interfaces where you swapped peer and message, revert modifications on SMSMessageHandler and that line in SMSHandler and we're probably merging this. For strategy handling in preferences, you could create another pull request so that it can be discussed properly.

Commit:
Fixed compilation errors in other classes

Commit:
Added some tests.

RicDeZen commented:
Ok, somehow I missed the last comment from Luca and just did everything, if this won't pass we'll just do another request with only the argument switch.

CremaLuca commented on smslibrary/src/main/java/com/eis/smslibrary/DefaultSMSParseStrategy.java:
```diff
@@ -0,0 +1,35 @@
+package com.eis.smslibrary;
+
+import androidx.annotation.NonNull;
+
```
A class description would fit well, like "Default parsing strategy used in {@link SMSMessageParser}, uses hidden character etc..."

CremaLuca commented on smslibrary/src/main/java/com/eis/smslibrary/SMSMessageParser.java:
```diff
@@ -0,0 +1,124 @@
+     * @param context The calling context
+     * @return {@link SMSMessageParser#instance} the only active instance of this class
+     */
+    public static SMSMessageParser getInstance(Context context){
```
Well thought, in this particular case passing a parameter in `getInstance` is better than having a `setup` method because you can only use two other methods that both need the `context`

CremaLuca commented on smslibrary/src/main/java/com/eis/smslibrary/SMSMessageParser.java:
```diff
@@ -0,0 +1,124 @@
+     * @param context the calling Context
+     * @param parseStrategy custom message parsing
+     */
+    public void setMessageParseStrategy(Context context, @NonNull final SMSParseStrategy parseStrategy) {
```
No need for context here, you already asked it in the constructor/getInstance, it's not good looking to have `SMSMessageeParser.getInstance(ctx).setMessageParseStrategy(ctx /* again */, customStrategy)`. Save it in the constructor.

CremaLuca commented on smslibrary/src/main/java/com/eis/smslibrary/SMSMessageParser.java:
```diff
@@ -0,0 +1,124 @@
+     * @param context the calling Context.
+     * @return The built SMSParseStrategy object, null if no valid class or constructor is found.
+     */
+    public SMSParseStrategy getSavedMessageParseStrategy(Context context){
```
As before.
Don't forget to remove it from the specifications.

CremaLuca replied:
I don't think the name fits well for a public method:
I think that if you were to keep it public you should rename it `getMessageParseStrategy` or `getCurrent...`.
Or you'd have to make it private but I'd keep it public, might be useful to someone or in tests.

CremaLuca commented on smslibrary/src/main/java/com/eis/smslibrary/SMSMessageParser.java:
```diff
@@ -0,0 +1,124 @@
+ * Uses a strategy to parse messages, so that any user can update it to its preferred parser
+ * By defaults it uses a default strategy, defined by the library
+ *
+ * @author Luca Crema, Alberto Ursino, Marco Mariotto
```
I can't see your name here! Get some credits, you deserve them

CremaLuca commented on smslibrary/src/main/java/com/eis/smslibrary/SMSParseStrategy.java:
```diff
@@ -0,0 +1,9 @@
+/**
+ * @author Riccardo De Zen
+ */
+public interface SMSParseStrategy extends MessageParseStrategy<String, SMSPeer, SMSMessage> {
```
Nice, much cleaner than before

CremaLuca commented on smslibrary/src/test/java/com/eis/smslibrary/TestSMSParseStrategy.java:
```diff
@@ -0,0 +1,34 @@
+ */
+class TestSMSParseStrategy implements SMSParseStrategy {
+
+    static final String KEY = "If violence is not the answer, the question itself is wrong.";
```
Where did you get this from? Can't find any reference on the internet.

RicDeZen replied:
I read it in a Berserk related Facebook group a long time ago, I guess it's a parody of other similar quotes because I never managed to find the source myself

CremaLuca reviewed (changes requested):
Noice, I'd approve it right away if it wasn't for a few things

CremaLuca commented:
ALSO, I forgot to mention that you should update the readme.MD file, especially the last part about strategies

Cogno-IDU commented on smslibrary/src/main/java/com/eis/smslibrary/SMSMessageParser.java:
```diff
@@ -0,0 +1,124 @@
+     */
+    public static SMSMessageParser getInstance(Context context){
+        if(instance == null)
+            return new SMSMessageParser(context);
```
Careful here! Since you're directly returning a new object, instance never gets setup, so you'll always return new objects!

RicDeZen replied:
Missed it, thanks

Cogno-IDU commented on smslibrary/src/main/java/com/eis/smslibrary/SMSMessageParser.java:
```diff
@@ -0,0 +1,124 @@
+     * Private constructor
+     * @param context The calling context
+     */
+    private SMSMessageParser(Context context) {
```
I would also check if the context is null here, throwing an exception, just in case.

Cogno-Marco reviewed (changes requested):
Requesting a singleton fix, the rest is perfect 👍

Commit:
Edits following review by Crema and Cognolato.

- Context is now saved in a WeakReference and has been removed from method parameters
- Methods operating with the Context check if it is null, and return false or null if so

Commit:
Fixed method calls in tests.

RicDeZen commented:
Suggested changes have been committed and pushed.

Cogno-Marco reviewed (approved):
Nicely Done

Commit:
Minor edits to README.md

giovannivelludo commented on smslibrary/src/main/java/com/eis/smslibrary/SMSHandler.java:
```diff
@@ -196,12 +196,12 @@ public void removeReceivedListener() {
      */
     private String getSMSContent(SMSMessage message) {
-        return SMSMessageHandler.getInstance().parseData(message);
+        return SMSMessageParser.getInstance(context.get()).parseData(message);
```
Here you're adding a character to the text of the `message` to be sent (you're calling method `parseData()` from `DefaultSMSPartStrategy`), but you're not checking if it still fits `SMSMessage.MAX_MSG_TEXT_LEN`

Cogno-IDU commented on smslibrary/src/main/java/com/eis/smslibrary/SMSMessageParser.java:
```diff
@@ -0,0 +1,134 @@
+     * @return {@link SMSMessageParser#instance} the only active instance of this class
+     */
+    public static SMSMessageParser getInstance(@NonNull Context callingContext){
+        if(callingContext == null) throw new NullPointerException(NULL_ERROR_MSG);
```
NonNull doesn't guarantee that it won't be null, it will just launch a warning in writing/compile time.

Cogno-IDU commented on smslibrary/src/main/java/com/eis/smslibrary/SMSMessageParser.java:
```diff
@@ -0,0 +1,134 @@
+    }
+
+    /**
+     * Interprets a string arrived via the communication channel and parses it to a library {@link SMSMessage}
```
"Parses it to a library SMSMessage" but there's a link to the last name

giovannivelludo replied:
thanks, the word "library" could be removed

RicDeZen commented:
I'll be closing this pull request since I forgot it even was still here and since it would just be faster to just reopen it later
